### ISSUE TITLE
Tensorflow Bitcast - float32 to int32 to float32?

### ISSUE THREAD

**Issue by etolliver** [tensorflow/tensorflow#30063]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: 
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): ('v1.10.1-0-g4dcfddc5d1', '1.10.1')
- Python version: 2.7.12
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CPU
- GPU model and memory: 

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Based off of the network, found at, 
https://www.easy-tensorflow.com/tf-tutorials/convolutional-neural-nets-cnns/cnn1
and after some additional modifications, I added layers before and after each convolutional layer that bitcasts the initial tf.float32 to a tf.int32 and back to a tf.float32. This resulted in a average accuracy loss of 5% compared to passing the tensor through the layer unmodified 

**Describe the expected behavior**
Based off of the documentation:
> tf.bitcast ( input, type, name=None )
> Given a tensor _input_, this operation returns a tensor that has the same buffer data as _input_ with datatype _type_.

I expected that a change and return to the original data type, without modifying the buffer data would not cause that amount of decrease to the network's accuracy.  

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

`
import tensorflow as tf
import numpy as np

old_v = tf.logging.get_verbosity()
tf.logging.set_verbosity(tf.logging.ERROR)

from tensorflow.examples.tutorials.mnist import input_data

#Image / Class Information
img_h = img_w = 28  # MNIST images are 28x28
img_size_flat = img_h * img_w  # 28x28=784, the total number of pixels
n_classes = 10  # Number of classes, one class per digit
n_channels = 1

 #Numpy Number Type
npNumberType = np.float32

def load_data(mode='train'):
    mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)
    if mode == 'train':
        x_train, y_train, x_valid, y_valid = mnist.train.images, mnist.train.labels, \
                                             mnist.validation.images, mnist.validation.labels
        x_train, _ = reformat(x_train, y_train)
        x_valid, _ = reformat(x_valid, y_valid)
        return x_train, y_train, x_valid, y_valid
    elif mode == 'test':
        x_test, y_test = mnist.test.images, mnist.test.labels
        x_test, _ = reformat(x_test, y_test)
    return x_test, y_test

def reformat(x, y):
    img_size, num_ch, num_class = int(np.sqrt(x.shape[-1])), 1, len(np.unique(np.argmax(y, 1)))
    dataset = x.reshape((-1, img_size, img_size, num_ch)).astype(npNumberType)
    labels = (np.arange(num_class) == y[:, None]).astype(npNumberType)
    return dataset, labels

def randomize(x, y):
    permutation = np.random.permutation(y.shape[0])
    shuffled_x = x[permutation, :, :, :]
    shuffled_y = y[permutation]
    return shuffled_x, shuffled_y

def get_next_batch(x, y, start, end):
    x_batch = x[start:end]
    y_batch = y[start:end]
    return x_batch, y_batch

x_train, y_train, x_valid, y_valid = load_data(mode='train')
print("Size of:")
print("- Training-set:\t\t{}".format(len(y_train)))
print("- Validation-set:\t{}".format(len(y_valid)))

def padding (x, name):
    dtypecast = tf.bitcast(x, tf.int32)
    return tf.bitcast(dtypecast, tf.float32, name=name)

#TensorFlow Number Format
tfGraphNumberType = tf.float32
rngseed = 1234

for runs in range(1, 2, 1):
    tf.reset_default_graph()
    tf.set_random_seed(rngseed)
    np.random.seed(rngseed)

    with tf.name_scope('Input'):
        x = tf.placeholder(tfGraphNumberType, shape=[None, img_h, img_w, n_channels], name='X')
        y = tf.placeholder(tfGraphNumberType, shape=[None, n_classes], name='Y')

    padding0 = padding(x, 'Pad0')

    # 1st Convolutional Layer
    filter_size1 = 5  # Convolution filters are 5 x 5 pixels.
    num_filters1 = 16  # There are 16 of these filters.
    stride1 = 1  # The stride of the sliding window

    with tf.variable_scope('conv1'):
        num_in_channel = padding0.get_shape().as_list()[-1]
        W = tf.get_variable('W', dtype=tfGraphNumberType,
                            shape=[filter_size1, filter_size1, num_in_channel, num_filters1],
                            initializer=tf.truncated_normal_initializer(stddev=0.01,seed=rngseed))
        tf.summary.histogram('weight', W)
        b = tf.get_variable('b', dtype=tfGraphNumberType,
                            initializer=tf.constant(0., shape=[num_filters1], dtype=tfGraphNumberType))
        tf.summary.histogram('bias', b)
        layer = tf.nn.conv2d(padding0, W, strides=[1, stride1, stride1, 1], padding="SAME")
        layer += b
        conv1 = tf.nn.relu(layer)

    shape = conv1.get_shape().as_list()[-1]
    padding1 = padding(conv1, 'Pad1')
    pool1 = tf.nn.max_pool(padding1, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding="SAME", name='pool1')

    # 2nd Convolutional Layer
    filter_size2 = 5  # Convolution filters are 5 x 5 pixels.
    num_filters2 = 32  # There are 32 of these filters.
    stride2 = 1  # The stride of the sliding window

    with tf.variable_scope('conv2'):
        num_in_channel = pool1.get_shape().as_list()[-1]
        W = tf.get_variable('W', dtype=tfGraphNumberType,
                            shape=[filter_size2, filter_size2, num_in_channel, num_filters2],
                            initializer=tf.truncated_normal_initializer(stddev=0.01, seed=rngseed))
        tf.summary.histogram('weight', W)
        b = tf.get_variable('b', dtype=tfGraphNumberType,
                            initializer=tf.constant(0., shape=[num_filters2], dtype=tfGraphNumberType))
        tf.summary.histogram('bias', b)
        layer = tf.nn.conv2d(pool1, W, strides=[1, stride2, stride2, 1], padding="SAME")
        layer += b
        conv2 = tf.nn.relu(layer)

    padding2 = padding(conv2, 'Pad2')
    pool2 = tf.nn.max_pool(padding2, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding="SAME", name='pool2')

    with tf.variable_scope('Flatten_layer'):
        layer_shape = pool2.get_shape()
        num_features = layer_shape[1:4].num_elements()
        layer_flat = tf.reshape(pool2, [-1, num_features])

    # Fully-connected layer.
    h1 = 128  # Number of neurons in fully-connected layer.

    with tf.variable_scope('FC1'):
        in_dim = layer_flat.get_shape()[1]
        W = tf.get_variable('W', dtype=tfGraphNumberType,
                            shape=[in_dim, h1],
                            initializer=tf.truncated_normal_initializer(stddev=0.01, seed=1234))
        tf.summary.histogram('weight', W)
        b = tf.get_variable('b', dtype=tfGraphNumberType,
                            initializer=tf.constant(0., shape=[h1], dtype=tfGraphNumberType))
        tf.summary.histogram('bias', b)

        layer = tf.matmul(layer_flat, W)
        layer += b
        fc1 = tf.nn.relu(layer)

    with tf.variable_scope('OUT'):
        in_dim = fc1.get_shape()[1]
        W = tf.get_variable('W', dtype=tfGraphNumberType,
                            shape=[in_dim, n_classes],
                            initializer=tf.truncated_normal_initializer(stddev=0.01, seed=rngseed))
        tf.summary.histogram('weight', W)
        b = tf.get_variable('b', dtype=tfGraphNumberType,
                            initializer=tf.constant(0., shape=[n_classes], dtype=tfGraphNumberType))
        tf.summary.histogram('bias', b)

        output_logits = tf.matmul(fc1, W)
        output_logits += b

    # Training Variables
    logs_path = "./logs"  # path to the folder that we want to save the logs for Tensorboard
    lr = 0.001  # The optimization initial learning rate
    epochs = 5  # Total number of training epochs
    batch_size = 100  # Training batch size
    display_freq = 50  # Frequency of displaying the training results
    tfOutputNumberType = tf.float32

    with tf.variable_scope('Train'):
        with tf.variable_scope('Loss'):
            loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y, logits=output_logits), name='loss')
        tf.summary.scalar('loss', loss)
        with tf.variable_scope('Optimizer'):
            optimizer = tf.train.AdamOptimizer(learning_rate=lr, name='Adam-op').minimize(loss)
        with tf.variable_scope('Accuracy'):
            correct_prediction = tf.equal(tf.argmax(output_logits, 1), tf.argmax(y, 1), name='correct_pred')
            accuracy = tf.reduce_mean(tf.cast(correct_prediction, tfOutputNumberType), name='accuracy')
        tf.summary.scalar('accuracy', accuracy)
        with tf.variable_scope('Prediction'):
            cls_prediction = tf.argmax(output_logits, axis=1, name='predictions')

    # Number of training iterations in each epoch
    num_tr_iter = int(len(y_train) / batch_size)

    # Initialize the variables
    init = tf.global_variables_initializer()
    # Merge all summaries
    merged = tf.summary.merge_all()

    sess = tf.InteractiveSession()
    sess.run(init)
    global_step = 0
    summary_writer = tf.summary.FileWriter(logs_path, sess.graph)

    print('R: {}'.format(runs))
    for epoch in range(epochs):
        #print('Training epoch: {}'.format(epoch + 1))
        x_train, y_train = randomize(x_train, y_train)
        for iteration in range(num_tr_iter):
            global_step += 1
            start = iteration * batch_size
            end = (iteration + 1) * batch_size
            x_batch, y_batch = get_next_batch(x_train, y_train, start, end)

            # Run optimization op (backprop)
            feed_dict_batch = {x: x_batch, y: y_batch}
            sess.run(optimizer, feed_dict=feed_dict_batch)

            if iteration % display_freq == 0:
                # Calculate and display the batch loss and accuracy
                loss_batch, acc_batch, summary_tr = sess.run([loss, accuracy, merged],
                                                             feed_dict=feed_dict_batch)
                summary_writer.add_summary(summary_tr, global_step)

                print("I:{0:3d}:\t L={1:.2f},\tA={2:.01%}".
                      format(iteration, loss_batch, acc_batch))

        # Run validation after every epoch
        feed_dict_valid = {x: x_valid, y: y_valid}
        loss_valid, acc_valid = sess.run([loss, accuracy], feed_dict=feed_dict_valid)

        print("M:{0}\tE:{1}\tF\tA:{2:.01%}".
              format(runs, epoch + 1, acc_valid))

    # Test the network when training is done
    x_test, y_test = load_data(mode='test')
    feed_dict_test = {x: x_test, y: y_test}
    loss_test, acc_test = sess.run([loss, accuracy], feed_dict=feed_dict_test)
    print('---------------------------------------------------------')
    print("Test loss: {0:.2f}, test accuracy: {1:.01%}".format(loss_test, acc_test))
    print('---------------------------------------------------------')

    # close the session after you are done with testing
    sess.close()
`

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

I attached a spreadsheet file with the results of training with different datatypes and my expectations. I apoplgize for any mistakes or missing information. Thank you.
[Bitcast Logs.xlsx](https://github.com/tensorflow/tensorflow/files/3318652/Bitcast.Logs.xlsx)

**Comment by achandraa** [comment]
I have tried in Colab with TF version 1.10.1 and was able to reproduce the issue.

**Comment by alextp** [comment]
Is this because bitcasting through integers generates not differentiable operations?

**Comment by gadagashwini** [comment]
Hi @etolliver, Note: `Bitcast` is implemented as a low-level cast, so machines with different endian orderings will give different results. Could you  try to run on colab with latest Tensorflow version. Thank you!

**Comment by etolliver** [comment]
Apologies,

I have been trying to find the original code I had back then. I will try to
reply asap
